{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Trying Unsupervised binning\n",
    "Metabat2 version 2.15 bioconda <br>\n",
    "concoct v1.1.0 bioconda <br>\n",
    "MaxBin 2.2.7 bioconda <br>\n",
    "binsanity <br>\n",
    "dastool 1.1.3 <br>\n",
    "<br>\n",
    "all the attempts etc was my trial and error to find the best binning parameter for everything. Once found the best one, binninmethod01 uses it.<br>\n",
    "\n",
    "MetaBCC-LR version 1.0.0 <br> <br>\n",
    "We want high completeness, low contamination https://github.com/Ecogenomics/CheckM/issues/65 <br>\n",
    "MAGs:\n",
    "- contamination < 10% <br>\n",
    "- completeness-5\\*contamination > 50 <br>\n",
    "The SH index is worth considering, but isn't nearly as critical as the estimated percentage of contamination. If the SH index is high (ideally 100%), it suggest the majority of contamination is from very similar species and thus any contamination is likely from the pangenome of the species being considered. Alternatively, if the SH index is very low (ideally 0%) this indicates all the contamination is likely from other species (perhaps very divergent species) and thus you may be able to identify it and remove it from the genome.\n",
    "\n",
    "then https://merenlab.org/2016/06/22/anvio-tutorial-v2/#anvi-import-collection\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metabat2\n",
    "https://bitbucket.org/berkeleylab/metabat/src/master/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "conda activate binning\n",
    "\n",
    "mkdir 05_binning\n",
    "mkdir 05_binning/metabat2\n",
    "mkdir 07_checkm\n",
    "\n",
    "for bam_sorted in $(find ./04_mapped_samples/ -maxdepth 1 -name \"*_sorted.bam\" -type f); do\n",
    "    file_name=$(basename $bam_sorted)\n",
    "    sample_name=${file_name%_*}\n",
    "    \n",
    "    # get contig depths\n",
    "    echo $sample_name\n",
    "    jgi_summarize_bam_contig_depths --outputDepth 05_binning/metabat2/${sample_name}_depth.txt $bam_sorted\n",
    "done\n",
    "\n",
    "\n",
    "for depths in $(find ./05_binning/metabat2/ -name \"*_depth.txt\" -type f); do\n",
    "    file_name=$(basename $depths)\n",
    "    sample_name=${file_name%_*}\n",
    "    sample_name=${sample_name%_*}\n",
    "    \n",
    "    #make folder\n",
    "    attempt=\"metabat_$sample_name\"\n",
    "    out_dir=\"05_binning/metabat2/${attempt}\"\n",
    "    mkdir $out_dir\n",
    "\n",
    "    #output fasta\n",
    "    metabat2 -i 03_correction_medaka/${sample_name}/consensus.fasta \\\n",
    "        -a $depths \\\n",
    "        -o $out_dir/bin \\\n",
    "        -t 32 --seed 1 --saveCls --verbose \\\n",
    "        --maxP 97 --minS 75\n",
    "\n",
    "    #output list\n",
    "    metabat2 -i 03_correction_medaka/${sample_name}/consensus.fasta \\\n",
    "        -a $depths \\\n",
    "        -o $out_dir/bin_fa_list \\\n",
    "        -t 32 --seed 1 -l \\\n",
    "        --maxP 97 --minS 75\n",
    "\n",
    "    #do checkm\n",
    "    checkm lineage_wf -x fa -t 32 --pplacer_threads 32 $out_dir 07_checkm/${attempt}\n",
    "    checkm qa 07_checkm/${attempt}/lineage.ms 07_checkm/${attempt} --tab_table > 07_checkm/${attempt}/results.txt\n",
    "done\n",
    "\n",
    "for depths in $(find ./05_binning/metabat2/ -name \"*_depth.txt\" -type f); do\n",
    "    file_name=$(basename $depths)\n",
    "    sample_name=${file_name%_*}\n",
    "    sample_name=${sample_name%_*}\n",
    "    attempt=\"metabat_$sample_name\"\n",
    "\n",
    "    checkm qa 07_checkm/${attempt}/lineage.ms 07_checkm/${attempt} --tab_table > 07_checkm/${attempt}/results.txt\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#metabat02 to dastool tab separated file\n",
    "R\n",
    "\n",
    "library(\"dplyr\")\n",
    "\n",
    "path_list <- dir(\"./05_binning/metabat2\", pattern = \"metabat\", full.names = T)\n",
    "path_save <- \"./05_binning/\"\n",
    "\n",
    "for (path in path_list){\n",
    "\n",
    "  bins_list <- list.files(path = path, pattern = \"bin_fa_list\", full.names = T)\n",
    "\n",
    "  bins_contigs <- data.frame(contigs = character(), bins = character())\n",
    "\n",
    "  for (i in 1:length(bins_list)){\n",
    "    bins_file_temp <- bins_list[i]\n",
    "    bin_no_temp <- strsplit(bins_file_temp, split = \"\\\\.\")[[1]][3]\n",
    "    bin_name_temp <- paste0(\"bin_\",bin_no_temp)\n",
    "    bins_list_temp <- read.table(bins_file_temp, stringsAsFactors = F)\n",
    "    bins_contigs_temp <- data.frame(contigs = bins_list_temp$V1, bins = bin_name_temp, stringsAsFactors = F)\n",
    "  \n",
    "    #append\n",
    "    bins_contigs <- bins_contigs %>% \n",
    "     rbind(bins_contigs_temp)\n",
    "  }\n",
    "\n",
    "  bins_contigs <- bins_contigs %>% \n",
    "    arrange(contigs)\n",
    "\n",
    "  write.table(bins_contigs, file = paste0(path_save, basename(path), \".txt\"), sep = \"\\t\", col.names = F, quote = F, row.names = F)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concoct\n",
    "https://concoct.readthedocs.io/en/latest/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "conda activate binning\n",
    "mkdir 05_binning/concoct\n",
    "\n",
    "for corr_folder in $(find ./03_correction_medaka/ -name \"sample*\" -type d); do\n",
    "    sample_name=$(basename $corr_folder)\n",
    "    out_dir=\"./05_binning/concoct\"\n",
    "    out_cut_fa=\"./05_binning/concoct/contigs_${sample_name}_10K\"\n",
    "    \n",
    "    # get cut up fasta\n",
    "    echo $sample_name\n",
    "    cut_up_fasta.py $corr_folder/consensus.fasta -c 10000 -o 0 --merge_last -b $out_cut_fa.bed > $out_cut_fa.fa\n",
    "    concoct_coverage_table.py $out_cut_fa.bed ./04_mapped_samples/${sample_name}_sorted.bam > $out_dir/coverage_table_${sample_name}.tsv \n",
    "    \n",
    "    # setup output directory for concoct results\n",
    "    attempt=\"concoct_$sample_name\"\n",
    "    out_dir_samp=\"05_binning/concoct/$attempt\"\n",
    "    mkdir $out_dir_samp\n",
    "\n",
    "    # run concoct\n",
    "    concoct --composition_file $out_cut_fa.fa \\\n",
    "        --coverage_file $out_dir/coverage_table_${sample_name}.tsv \\\n",
    "        -b $out_dir_samp -t 16 -s 1 \\\n",
    "        --total_percentage_pca 99 -c 800\n",
    "\n",
    "    #merge subcontig clustering\n",
    "    merge_cutup_clustering.py $out_dir_samp/clustering_gt1000.csv > $out_dir_samp/clustering_merged.csv \n",
    "\n",
    "    #export fasta\n",
    "    mkdir $out_dir_samp/fasta_bins\n",
    "    extract_fasta_bins.py $corr_folder/consensus.fasta \\\n",
    "        $out_dir_samp/clustering_merged.csv \\\n",
    "        --output_path $out_dir_samp/fasta_bins\n",
    "\n",
    "    #Delete all bins sized less than 500kb\n",
    "    cp -R $out_dir_samp/fasta_bins $out_dir_samp/fasta_bins_FINAL\n",
    "    find $out_dir_samp/fasta_bins_FINAL -name \"*.fa\" -type f -size -500k -delete\n",
    "    #now only have 244 bins phew\n",
    "\n",
    "    #checkm\n",
    "    checkm lineage_wf -x fa -t 16 --pplacer_threads 16 \\\n",
    "        $out_dir_samp/fasta_bins_FINAL \\\n",
    "        07_checkm/${attempt}\n",
    "    checkm qa 07_checkm/${attempt}/lineage.ms 07_checkm/${attempt} --tab_table > 07_checkm/${attempt}/results.txt\n",
    "done\n",
    "\n",
    "for corr_folder in $(find ./03_correction_medaka/ -name \"sample*\" -type d); do\n",
    "    sample_name=$(basename $corr_folder)\n",
    "    # setup output directory for concoct results\n",
    "    attempt=\"concoct_$sample_name\"\n",
    "    out_dir_samp=\"05_binning/concoct/$attempt\"\n",
    "    mkdir $out_dir_samp\n",
    "\n",
    "\n",
    "    #Delete all bins sized less than 500kb\n",
    "    cp -R $out_dir_samp/fasta_bins $out_dir_samp/fasta_bins_FINAL\n",
    "    find $out_dir_samp/fasta_bins_FINAL -name \"*.fa\" -type f -size -500k -delete\n",
    "    #now only have 244 bins phew\n",
    "\n",
    "    #checkm\n",
    "    checkm lineage_wf -x fa -t 16 --pplacer_threads 16 \\\n",
    "        $out_dir_samp/fasta_bins_FINAL \\\n",
    "        07_checkm/${attempt}\n",
    "    checkm qa 07_checkm/${attempt}/lineage.ms 07_checkm/${attempt} --tab_table > 07_checkm/${attempt}/results.txt\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "concoct01 = --total_percentage_pca 99 still contaminated liao. But best so far<br>\n",
    "concoct02 = --total_percentage_pca 99 -c 800<br>\n",
    "concoct05 = --total_percentage_pca 99 -i 750 same as 02 <br>\n",
    "concoct06 = --total_percentage_pca 99 -c 1000<br>\n",
    "concoct07 = --total_percentage_pca 99 -c 5000 not useful, stick with 02<br> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "R\n",
    "\n",
    "library(\"dplyr\")\n",
    "\n",
    "path_list <- dir(\"./05_binning/concoct\", pattern = \"concoct_sample\", full.names = T)\n",
    "path_save <- \"./05_binning/\"\n",
    "\n",
    "for (path in path_list){\n",
    "  bins_list <- list.files(path = path, pattern = \"bin_fa_list\", full.names = T)\n",
    "  bins_contigs <- data.frame(contigs = character(), bins = character())\n",
    "\n",
    "  bins_contigs <- read.csv(paste0(path,\"/clustering_merged.csv\"), stringsAsFactors = F)\n",
    "  bins <- paste0(\"bin_\",bins_contigs$cluster_id)\n",
    "  bins_contigs <- bins_contigs %>% \n",
    "    mutate(bins = bins) %>% \n",
    "    select(-cluster_id) %>% \n",
    "    arrange(contig_id)\n",
    "\n",
    "  write.table(bins_contigs, file = paste0(path_save, basename(path), \".txt\"), sep = \"\\t\", col.names = F, quote = F, row.names = F)\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Maxbin\n",
    "see the txt file that you downloaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "mkdir 05_binning/maxbin\n",
    "\n",
    "# Get depth txt\n",
    "\n",
    "R\n",
    "\n",
    "library(\"dplyr\")\n",
    "depth_files <- list.files(\"./05_binning/metabat2/\", pattern = \"depth\", full.names = TRUE)\n",
    "\n",
    "for (i in depth_files){\n",
    "  depth <- read.delim(i, stringsAsFactors = F)\n",
    "  depth <- depth %>% \n",
    "    select(-contains(\".var\")) %>% \n",
    "    select(-contains(\"Depth\")) %>% \n",
    "    select(-contains(\"Len\"))\n",
    "\n",
    "  for (i in 2:ncol(depth)){\n",
    "    depth_temp <- depth[,c(1, i)]\n",
    "    depth_name <- colnames(depth_temp)[2]\n",
    "    depth_name <- strsplit(depth_name, \"_\")[[1]]\n",
    "    depth_name <- depth_name[1]\n",
    "    depth_name <- paste0(depth_name,\"_depth.txt\")\n",
    "    write.table(depth_temp, file = paste0(\"05_binning/maxbin/\", depth_name), sep = \"\\t\", col.names = F, quote = F, row.names = F)\n",
    "  }\n",
    "}\n",
    "\n",
    "# MaxBin\n",
    "for corr_folder in $(find ./03_correction_medaka/ -name \"sample*\" -type d); do\n",
    "    sample_name=$(basename $corr_folder)\n",
    "    attempt=\"maxbin_${sample_name}\"\n",
    "    mkdir 05_binning/maxbin/${attempt}\n",
    "\n",
    "    run_MaxBin.pl \\\n",
    "        -contig $corr_folder/consensus.fasta \\\n",
    "        -out 05_binning/maxbin/${attempt}/bin \\\n",
    "        -thread 16 \\\n",
    "        -abund 05_binning/maxbin/${sample_name}_depth.txt \\\n",
    "        -prob_threshold 0.99\n",
    "\n",
    "    checkm lineage_wf -x fasta -t 16 --pplacer_threads 16 \\\n",
    "        05_binning/maxbin/${attempt}/ \\\n",
    "        07_checkm/${attempt}\n",
    "    checkm qa 07_checkm/${attempt}/lineage.ms 07_checkm/${attempt} --tab_table > 07_checkm/${attempt}/results.txt\n",
    "\n",
    "    # Get bins list\n",
    "    outdir=\"./05_binning/maxbin/${attempt}/bin_deets\"\n",
    "    mkdir $outdir\n",
    "\n",
    "    for bins in $(find ./05_binning/maxbin/${attempt}/ -maxdepth 1 -name \"*.fasta\" -type f); do\n",
    "        bin_name=$(basename $bins)\n",
    "        bin_txt=$outdir/${bin_name}.txt\n",
    "        grep \"^>\" $bins > $bin_txt\n",
    "    done\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "maxbin01: -prob_threshold 0.98 <br>\n",
    "maxbin02: -prob_threshold 0.98 -markerset 40 PAKE INI AJA DEH<br>\n",
    "maxbin03: -prob_threshold 0.99 <br>\n",
    "maxbin04: -prob_threshold 0.99 -markerset 40 <br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#THE BELOW IN R\n",
    "R\n",
    "\n",
    "library(\"dplyr\")\n",
    "\n",
    "path_list <- dir(\"./05_binning/maxbin\", pattern = \"maxbin_sample\", full.names = T)\n",
    "path_save <- \"./05_binning/\"\n",
    "\n",
    "for (path in path_list){\n",
    "  bins_list <- list.files(path = paste0(path,\"/bin_deets\"), pattern = \".fasta.txt\", full.names = T)\n",
    "  bins_contigs <- data.frame(contigs = character(), bins = character())  \n",
    "\n",
    "\n",
    "  for (i in 1:length(bins_list)){\n",
    "  bins_file_temp <- bins_list[i]\n",
    "  bin_no_temp <- as.numeric(strsplit(bins_file_temp, split = \"\\\\.\")[[1]][3])\n",
    "  bin_name_temp <- paste0(\"bin_\",bin_no_temp)\n",
    "  bins_list_temp <- read.table(bins_file_temp, stringsAsFactors = F)\n",
    "  bins_list_temp <- sub('.', '', bins_list_temp$V1)\n",
    "  \n",
    "  bins_contigs_temp <- data.frame(contigs = bins_list_temp, bins = bin_name_temp, stringsAsFactors = F)\n",
    "  \n",
    "  #append\n",
    "  bins_contigs <- bins_contigs %>% \n",
    "    rbind(bins_contigs_temp)\n",
    "}\n",
    "\n",
    "  bins_contigs <- bins_contigs %>% \n",
    "    arrange(contigs)\n",
    "\n",
    "  write.table(bins_contigs, file = paste0(path_save, basename(path), \".txt\"), sep = \"\\t\", col.names = F, quote = F, row.names = F)\n",
    "}\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Binsanity\n",
    "https://github.com/edgraham/BinSanity/wiki/Usage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Binsanity\n",
    "mkdir binsanity_temp\n",
    "temp=\"./binsanity_temp\"\n",
    "\n",
    "for corr_folder in $(find ./03_correction_medaka/ -name \"sample*\" -type d); do\n",
    "    sample_name=$(basename $corr_folder)\n",
    "    out_dir=\"./05_binning/binsanity\"\n",
    "    attempt=\"binsanity_${sample_name}\"\n",
    "    \n",
    "    cp 04_mapped_samples/${sample_name}_sorted* ${temp}/\n",
    "\n",
    "    Binsanity-profile -i $corr_folder/consensus.fasta \\\n",
    "    -s $temp \\\n",
    "    -o $out_dir/ \\\n",
    "    -c $out_dir/${sample_name} \\\n",
    "    -T 32\n",
    "\n",
    "    mkdir 05_binning/binsanity/${attempt}\n",
    "    Binsanity-wf -f $corr_folder -l consensus.fasta \\\n",
    "        -c $out_dir/${sample_name}.cov.x100.lognorm \\\n",
    "        -o $out_dir/${attempt}/ \\\n",
    "        --threads 32 \\\n",
    "        -p -3 --refine-preference -1\n",
    "\n",
    "\n",
    "    #Got 556 bins wow. Delete all bins sized less than 500kb\n",
    "    cp -R $out_dir/${attempt}/BinSanity-Final-bins $out_dir/${attempt}/BinSanity-Final-bins-500k\n",
    "    find $out_dir/${attempt}/BinSanity-Final-bins-500k -name \"*.fna\" -type f -size -500k -delete\n",
    "    #now only have 244 bins phew\n",
    "\n",
    "    #CheckM\n",
    "    checkm lineage_wf -x fna -t 16 --pplacer_threads 16 \\\n",
    "        $out_dir/${attempt}/BinSanity-Final-bins-500k \\\n",
    "        07_checkm/${attempt}/\n",
    "    checkm qa 07_checkm/${attempt}/lineage.ms 07_checkm/${attempt} --tab_table > 07_checkm/${attempt}/results.txt\n",
    "\n",
    "    #Prepare for R\n",
    "    mkdir $out_dir/${attempt}/BinSanity-Final-bins-500k/bin_deets\n",
    "    for bins in $(find ./$out_dir/${attempt}/BinSanity-Final-bins-500k/ -maxdepth 1 -name \"*.fna\" -type f); do\n",
    "        bin_name=$(basename $bins)\n",
    "        bin_txt=$out_dir/${attempt}/BinSanity-Final-bins-500k/bin_deets/${bin_name}.txt\n",
    "        grep \"^>\" $bins > $bin_txt\n",
    "    done\n",
    "    rm $temp/*\n",
    "done\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#buat indrik ganti\n",
    "mkdir binsanity_temp\n",
    "temp=\"./binsanity_temp\"\n",
    "\n",
    "for corr_folder in $(find ./03_correction_medaka/ -name \"sample*\" -type d); do\n",
    "    sample_name=$(basename $corr_folder)\n",
    "    out_dir=\"./05_binning/binsanity\"\n",
    "    attempt=\"binsanity_${sample_name}\"\n",
    "    \n",
    "    cp 04_mapped_samples/${sample_name}_sorted* ${temp}/\n",
    "\n",
    "    Binsanity-profile -i $corr_folder/consensus.fasta \\\n",
    "    -s $temp \\\n",
    "    -o $out_dir/ \\\n",
    "    -c $out_dir/${sample_name} \\\n",
    "    -T 32\n",
    "\n",
    "    mkdir 05_binning/binsanity/${attempt}\n",
    "    Binsanity-wf -f $corr_folder -l consensus.fasta \\\n",
    "        -c $out_dir/${sample_name}.cov.x100.lognorm \\\n",
    "        -o $out_dir/${attempt}/ \\\n",
    "        --threads 32 \\\n",
    "        -p -3 --refine-preference -1\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    #Got 556 bins wow. Delete all bins sized less than 500kb\n",
    "    cp -R $out_dir/${attempt}/BinSanity-Final-bins $out_dir/${attempt}/BinSanity-Final-bins-500k\n",
    "    find $out_dir/${attempt}/BinSanity-Final-bins-500k -name \"*.fna\" -type f -size -500k -delete\n",
    "    #now only have 244 bins phew\n",
    "\n",
    "    #CheckM\n",
    "    checkm lineage_wf -x fna -t 16 --pplacer_threads 16 \\\n",
    "        $out_dir/${attempt}/BinSanity-Final-bins-500k \\\n",
    "        07_checkm/${attempt}/\n",
    "    checkm qa 07_checkm/${attempt}/lineage.ms 07_checkm/${attempt} --tab_table > 07_checkm/${attempt}/results.txt\n",
    "\n",
    "    #Prepare for R\n",
    "    mkdir $out_dir/${attempt}/BinSanity-Final-bins-500k/bin_deets\n",
    "    for bins in $(find ./$out_dir/${attempt}/BinSanity-Final-bins-500k/ -maxdepth 1 -name \"*.fna\" -type f); do\n",
    "        bin_name=$(basename $bins)\n",
    "        bin_txt=$out_dir/${attempt}/BinSanity-Final-bins-500k/bin_deets/${bin_name}.txt\n",
    "        grep \"^>\" $bins > $bin_txt\n",
    "    done\n",
    "    rm $temp/*\n",
    "done\n",
    "\n",
    "# works normally\n",
    "Binsanity-wf -f ./ -l sample04_contigs.fasta \\\n",
    "        -c sample04.cov.x100.lognorm \\\n",
    "        -o ./test/ \\\n",
    "        --threads 32 \\\n",
    "        -p -3 --refine-preference -1\n",
    "\n",
    "# doesn't work\n",
    "Binsanity-wf -f ./ -l sample03_contigs.fasta \\\n",
    "        -c sample03.cov.x100.lognorm \\\n",
    "        -o ./test/ \\\n",
    "        --threads 32 \\\n",
    "        -p -3 --refine-preference -1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "binsanity01 = default. SO HIGH CONTAMINATION <br>\n",
    "binsanity02 = -p -3 --refine-preference -10 <br>\n",
    "binsanity03 = -p -3 --refine-preference -3 <br>\n",
    "binsanity04 = -p -3 --refine-preference -1 THIS IS THE BEST<br>\n",
    "binsanity05 = -p -1 --refine-preference -1 <br>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#THE BELOW IN R\n",
    "R\n",
    "\n",
    "library(\"dplyr\")\n",
    "\n",
    "path_list <- dir(\"./05_binning/binsanity\", pattern = \"binsanity_sample\", full.names = T)\n",
    "path_save <- \"./05_binning/\"\n",
    "\n",
    "for (path in path_list){\n",
    "  bins_list <- list.files(path = paste0(path, \"/BinSanity-Final-bins-500k/bin_deets/\"), pattern = \".txt\", full.names = T)\n",
    "  bins_contigs <- data.frame(contigs = character(), bins = character())\n",
    "\n",
    "  for (i in 1:length(bins_list)){\n",
    "    bins_file_temp <- bins_list[i]\n",
    "    bin_name_temp <- paste0(\"bin_\",i)\n",
    "    bins_list_temp <- read.table(bins_file_temp, stringsAsFactors = F)\n",
    "    bins_list_temp <- sub('.', '', bins_list_temp$V1)\n",
    "  \n",
    "    bins_contigs_temp <- data.frame(contigs = bins_list_temp, bins = bin_name_temp, stringsAsFactors = F)\n",
    "  \n",
    "    #append\n",
    "    bins_contigs <- bins_contigs %>% \n",
    "      rbind(bins_contigs_temp)\n",
    "  }\n",
    "\n",
    "  bins_contigs <- bins_contigs %>% \n",
    "    arrange(contigs)\n",
    "\n",
    "  write.table(bins_contigs, file = paste0(path_save, basename(path), \".txt\"), sep = \"\\t\", col.names = F, quote = F, row.names = F)\n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DASTOOL\n",
    "Fasta_to_Scaffolds2Bin.sh -e fasta > my_scaffolds2bin.tsv WAHLAO EH BUT THIS STILL DOESN'T WORK HAHA\n",
    "\n",
    "metabat02, maxbin02, binsanity04, concoct02"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir 05_binning/dastool/\n",
    "\n",
    "for corr_folder in $(find ./03_correction_medaka/ -name \"sample*\" -type d); do\n",
    "    sample_name=$(basename $corr_folder)\n",
    "    attempt=\"dastool_$sample_name\"\n",
    "    out_dir=\"./05_binning/dastool/$attempt\"\n",
    "    mkdir $out_dir\n",
    "\n",
    "    temp=$(find ./05_binning/ -name \"*${sample_name}.txt\" -type f | tr '\\n' '\\,')\n",
    "\n",
    "    echo \"doing DAS_Tool\"\n",
    "    echo $sample_name\n",
    "    echo $temp\n",
    "    echo $out_dir/$attempt\n",
    "\n",
    "    DAS_Tool \\\n",
    "        --bins $temp \\\n",
    "        --contigs $corr_folder/consensus.fasta \\\n",
    "        --outputbasename $out_dir/$attempt \\\n",
    "        --threads 32 --write_bins 1 \\\n",
    "        --search_engine diamond\n",
    "\n",
    "    checkm lineage_wf -x fa -t 32 --pplacer_threads 32 \\\n",
    "        $out_dir/${attempt}_DASTool_bins \\\n",
    "        07_checkm/${attempt}/\n",
    "    checkm qa 07_checkm/${attempt}/lineage.ms 07_checkm/${attempt} --tab_table > 07_checkm/${attempt}/results.txt\n",
    "\n",
    "    #replace \".\" with \"_\" in the bin list\n",
    "    tr '\\\\.' '_' < $out_dir/${attempt}_DASTool_scaffolds2bin.txt > $out_dir/${attempt}_DASTool_scaffolds2bin_underscore.txt\n",
    "    \n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#then import results to anvio for manual refining.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "# Collate all in FINAL bins\n",
    "mkdir 05_binning/aa_FINAL_01_all\n",
    "cp 05_binning/dastool/dastool_sample*/dastool_sample*_DASTool_bins/* 05_binning/aa_FINAL_01_all/\n",
    "find ./05_binning/aa_FINAL_01_all -name \"*-CLEAN*\" -type f -size -500k -delete\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove contigs without SCGs\n",
    "## Checking if contigs are correctly binned\n",
    "Checking for single-copy marker genes --> CheckM on fasta with 1 less contig. If contigs are correctly binned, then completion should be less, contamination stay same."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "mkdir 05_binning/aa_FINAL_02_TRIMMED\n",
    "mkdir 10_trimming\n",
    "mkdir 07_checkm/zz_Trimming\n",
    "conda activate binning\n",
    "\n",
    "\n",
    "for bin in $(find ./05_binning/aa_FINAL_01_all/ -name \"*.fa\" -type f); do\n",
    "    attempt=$(basename $bin)\n",
    "    attempt=${attempt%.*}\n",
    "    mkdir 10_trimming/$attempt\n",
    "    mkdir 10_trimming/${attempt}_contigs\n",
    "\n",
    "    # split multifasta into separate single fasta\n",
    "    # https://gist.github.com/astatham/621901\n",
    "    cat $bin | awk '{\n",
    "        if (substr($0, 1, 1)==\">\") {filename=(substr($0,2) \".fa\")}\n",
    "        print $0 > filename\n",
    "    }'\n",
    "    # move it to the contigs folder\n",
    "    mv Utg*.fa 10_trimming/${attempt}_contigs/\n",
    "\n",
    "    # sequentially cat all except one \n",
    "    for header in $(grep \">\" $bin); do\n",
    "        utg=\"${header:1}\"\n",
    "        cat $(find ./10_trimming/${attempt}_contigs/ -not -name \"*$utg*\" -type f) > 10_trimming/${attempt}/${attempt}_no_$utg.fa\n",
    "    done\n",
    "    cp $bin 10_trimming/${attempt}/${attempt}_no_deletions.fa\n",
    "\n",
    "    checkm lineage_wf -x fa -t 64 --pplacer_threads 64 \\\n",
    "        10_trimming/$attempt/ \\\n",
    "        07_checkm/zz_Trimming/${attempt}\n",
    "    checkm qa 07_checkm/zz_Trimming/${attempt}/lineage.ms 07_checkm/zz_Trimming/${attempt} --tab_table > 07_checkm/zz_Trimming/${attempt}/results.txt\n",
    "\n",
    "    # dRep editing checkM results\n",
    "    grep -v '\\[' 07_checkm/zz_Trimming/${attempt}/results.txt > 07_checkm/zz_Trimming/${attempt}/results_edit.txt\n",
    "    sed -i 's/Completeness/completeness/gI' 07_checkm/zz_Trimming/${attempt}/results_edit.txt\n",
    "    sed -i 's/Contamination/contamination/gI' 07_checkm/zz_Trimming/${attempt}/results_edit.txt\n",
    "    sed -i 's/\\t/\\,/gI' 07_checkm/zz_Trimming/${attempt}/results_edit.txt\n",
    "    awk 'BEGIN{FS=OFS=\",\"}{$1=$1\".fa\"}1' 07_checkm/zz_Trimming/${attempt}/results_edit.txt > 07_checkm/zz_Trimming/${attempt}/results_edit_genome.txt\n",
    "    sed -i 's/Bin\\ Id.fa/genome/gI' 07_checkm/zz_Trimming/${attempt}/results_edit_genome.txt\n",
    "    mv 07_checkm/zz_Trimming/${attempt}/results_edit_genome.txt 07_checkm/zz_Trimming/${attempt}/results_edit_genome.csv\n",
    "\n",
    "    Rscript get_core_contigs.R 07_checkm/zz_Trimming/${attempt}/results_edit_genome.csv\n",
    "\n",
    "    # separate contigs with SCGs and contigs without\n",
    "    input=\"07_checkm/zz_Trimming/${attempt}/core.txt\"\n",
    "    while IFS= read -r line\n",
    "    do\n",
    "        cat 10_trimming/${attempt}_contigs/$line >> 05_binning/aa_FINAL_02_TRIMMED/$attempt.fa\n",
    "    done < \"$input\"\n",
    "\n",
    "\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# dRep to dereplicate MAGs from bins of individual samples\n",
    "\n",
    "Checkm, GTDBTK, dRep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#CHECKM\n",
    "attempt=\"aa_FINAL_02_TRIMMED\"\n",
    "\n",
    "conda deactivate\n",
    "conda activate binning\n",
    "checkm lineage_wf -x fa -t 32 --pplacer_threads 32 \\\n",
    "        05_binning/${attempt}/ \\\n",
    "        07_checkm/${attempt}/\n",
    "checkm qa 07_checkm/${attempt}/lineage.ms 07_checkm/${attempt} --tab_table > 07_checkm/${attempt}/results.txt\n",
    "\n",
    "#GTDBTK2\n",
    "attempt=\"aa_FINAL_02_TRIMMED\"\n",
    "mkdir 06_taxa_gtdbtk2\n",
    "conda deactivate\n",
    "conda activate gtdbtk2\n",
    "\n",
    "gtdbtk classify_wf \\\n",
    "    --genome_dir 05_binning/${attempt} \\\n",
    "    --out_dir 06_taxa_gtdbtk2/${attempt}/ \\\n",
    "    -x fa --cpus 32\n",
    "\n",
    "# dRep editing checkM results\n",
    "grep -v '\\[' 07_checkm/${attempt}/results.txt > 07_checkm/${attempt}/results_edit.txt\n",
    "sed -i 's/Completeness/completeness/gI' 07_checkm/${attempt}/results_edit.txt\n",
    "sed -i 's/Contamination/contamination/gI' 07_checkm/${attempt}/results_edit.txt\n",
    "sed -i 's/\\t/\\,/gI' 07_checkm/${attempt}/results_edit.txt\n",
    "awk 'BEGIN{FS=OFS=\",\"}{$1=$1\".fa\"}1' 07_checkm/${attempt}/results_edit.txt > 07_checkm/${attempt}/results_edit_genome.txt\n",
    "sed -i 's/Bin\\ Id.fa/genome/gI' 07_checkm/${attempt}/results_edit_genome.txt\n",
    "mv 07_checkm/${attempt}/results_edit_genome.txt 07_checkm/${attempt}/results_edit_genome.csv\n",
    "\n",
    "# dRep!\n",
    "conda deactivate\n",
    "conda activate dRep_dependencies\n",
    "out_dir=\"./05_binning/aa_FINAL_03_TRIMMED_dRep_95\"\n",
    "mkdir $out_dir\n",
    "rm -r $out_dir/*\n",
    "dRep dereplicate $out_dir  -p 32 -g ./05_binning/${attempt}/*.fa --genomeInfo ./07_checkm/${attempt}/results_edit_genome.csv \n",
    "\n",
    "# add metabat_sample02.txt.007 because classified differently than all others in gtdbtk, has MQ assembly/binning, but not in dRep.\n",
    "mkdir ./05_binning/aa_FINAL_04_TRIMMED_CONFIRM\n",
    "cp ./05_binning/aa_FINAL_03_TRIMMED_dRep_95/dereplicated_genomes/* ./05_binning/aa_FINAL_04_TRIMMED_CONFIRM/\n",
    "cp ./05_binning/aa_FINAL_02_TRIMMED/metabat_sample02.txt.007.fa ./05_binning/aa_FINAL_04_TRIMMED_CONFIRM/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#CHECKM\n",
    "attempt=\"aa_FINAL_04_TRIMMED_CONFIRM\"\n",
    "\n",
    "conda deactivate\n",
    "conda activate binning\n",
    "checkm lineage_wf -x fa -t 32 --pplacer_threads 32 \\\n",
    "        05_binning/${attempt}/ \\\n",
    "        07_checkm/${attempt}/\n",
    "checkm qa 07_checkm/${attempt}/lineage.ms 07_checkm/${attempt} --tab_table > 07_checkm/${attempt}/results.txt\n",
    "\n",
    "#GTDBTK2\n",
    "attempt=\"aa_FINAL_04_TRIMMED_CONFIRM\"\n",
    "mkdir 06_taxa_gtdbtk2\n",
    "conda deactivate\n",
    "conda activate gtdbtk2\n",
    "\n",
    "gtdbtk classify_wf \\\n",
    "    --genome_dir 05_binning/${attempt} \\\n",
    "    --out_dir 06_taxa_gtdbtk2/${attempt}/ \\\n",
    "    -x fa --cpus 32\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
